51 the process of recording the type of physical event in g4

jhajduga opened the pull request.

jhajduga committed:
WIP

jhajduga committed:
TempChanges signed as "// TEMP!!!!"

jhajduga committed:
Updated:
More extensive list of processes to save within NTuple.
Added ability to record electron origin within NTuple.

jhajduga committed:
A major physiclist update for more accurate recording of the processes and reactions taking place inside the scintillator.

jhajduga committed:
 fix: merge run-level data using G4Cache - eliminating multiple dose counts

jhajduga committed:
Optimise simulation for new physic model.

jhajduga commented on core/physics/PhysicsList.cc:
```diff
@@ -57,34 +66,109 @@ void PhysicsList::AddStepMax()
-  m_decayPhysicsModelCtr->ConstructProcess();
+  // --- Mandatory transportation process ---
+  AddTransportation();  // Enables particles to move through geometry
+
```
Add m_decayPhysicsModelCtr->ConstructProcess();

jhajduga commented on core/physics/PhysicsList.cc:
```diff
@@ -57,34 +66,109 @@ void PhysicsList::AddStepMax()
+  // --- Electromagnetic physics processes (standard EM model) ---
+  m_emPhysicsModelCtr->ConstructProcess();  // Custom EM physics builder
+
+  // --- Additional EM processes (optical effects) ---
```
All code below - if flag then use it.
Otherwise - use standard em_standard physic.

jhajduga committed:
Update.

jhajduga committed:
Update gitignore.

barachwal commented on core/analysis/NTupleEventAnalisys.cc:
```diff
@@ -15,13 +15,17 @@
 #include <vector>
 #include <set>
 
+G4ThreadLocal NTupleEventAnalisys* NTupleEventAnalisys::fInstance = nullptr;
```
- [ ] To be removed

barachwal commented on core/analysis/NTupleEventAnalisys.hh:
```diff
@@ -142,5 +144,7 @@ class NTupleEventAnalisys {
     static const std::vector<TTreeCollection>& TreeCollection() { return m_ttree_collection.Get(); }
 
+    /// Thread-local singleton instance pointer.
+    G4ThreadLocal static NTupleEventAnalisys* fInstance;
```
- [ ] Remove G4ThreadLocal
- [ ] Use convention: fInstance -> m_instance;

barachwal commented on core/geometry/Patient/VoxelHit.cc:
```diff
@@ -42,7 +42,9 @@ void VoxelHit::Fill(G4Step* aStep){
-
+    
+    // Shouldn't we move the if here that sits inside FillTrack?
+    // Unnecessary going into the method.I don't know if it changes anything but... TODO:?
```
O co chodzi??? ewentualnie komentarze do usunięcia

jhajduga replied:
Wewnątrz funkcji:
void VoxelHit::FillTrack(G4Step* aStep){
Całość kodu jest w:
  if(m_tracks_analysis){ ... } 
  
Zastanawiałem się, czy tego warunku (  if(m_tracks_analysis) ) nie przenieść  w miejsce gdzie wywołujemy FillTrack, żeby niepotrzebnie nie wchodzić w tę funkcję (Byłem zaraz po przejrzeniu filmu o optymalizacji C++ i myślałem, że jest to róż nica kilku dobrych instrukcji assemblera która w całym scoopie naszego programu nic nie zmieni))

barachwal commented on core/geometry/Patient/VoxelHit.hh:
```diff
@@ -202,6 +204,8 @@ class VoxelHit final : public G4VHit {
   std::vector<std::pair<G4int,G4ThreeVector>> GetTrkPosition() const;
   std::vector<std::pair<G4int,G4int>> GetTrkType() const;
+  std::vector<std::pair<G4int,G4int>> GetProcessType() const;
+  std::vector<std::pair<G4int,G4int>> GetElectronOriginType() const;
```
- [x] Przy okazji tego PR, proponuje modyfikacje nazw tych funkcji:
GetTrkEnergy -> GetTrkIdEnergyMappingList 
GetTrkTheta -> GetTrkIdThetaMappingList 
GetTrkLength -> GetTrkIdLengthMappingList 
GetUserTrkLength -> GetTrkIdUserLengthMappingList 
GetTrkPosition -> GetTrkIdPositionMappingList 
GetTrkType -> GetTrackIdTypeMappingList
GetProcessType -> GetTrkIdProcessTypeMappingList
GetElectronOriginType -> GetTrkIdElectronOriginTypeMappingList

barachwal commented on core/geometry/Patient/WaterPhantom/WaterPhantom.cc:
```diff
@@ -144,7 +144,7 @@ void WaterPhantom::Construct(G4VPhysicalVolume *parentWorld) {
   auto regVol = new G4Region("waterPhantomR");
   auto cuts = new G4ProductionCuts;
-  cuts->SetProductionCut(2.0 * mm);
+  cuts->SetProductionCut(2.5 * mm);
```
Czy ta zmiana wynika z czegoś konkretnego? Warto dać komentarz w kodzie

jhajduga replied:
Temp zmiana przy testowaniu.

barachwal commented on core/physics/PhysicsList.cc:
```diff
@@ -57,34 +66,109 @@ void PhysicsList::AddStepMax()
+// }
+
+
+// TEMP?:
```
- [ ] clean up, remove commented code

barachwal commented on jobs/singleScintilator/Cez.toml:
```diff
@@ -0,0 +1,44 @@
+[RunSvc]
```
Nazwać plik po angielsku

jhajduga committed:
Wszystkie elementy symulacji działają, schemat nowej ifologi i mechanizm tej ifologi gotowe -> zostało tylko wprowadzić te ify odpowiednio i wprowadzić zmiany proponowane przez bartka.

jhajduga committed:
PhysicList redoo.
Analysis flag redoo.
Proposed changes implemented.

TODO:
- Usage of flags in "Scoring" eg:
  TracksAnalysis = true -> StoreTracks=true
  MinimalMode=true
  Itp.
- Rename func name in VoxelHit.

jhajduga committed:
Conditional criteria during tree creation -> done

TODO:
- Rename methods in Voxel Hit.
- Enable/Disable flags on basic of job definition.

jhajduga committed:
Renaming in voxelhit done.

TODO's for future defined.

Before merge TODO:
  - Enable/Disable flags on basic of job definition.

jhajduga commented:
Stil TODO:
 - Enable/Disable flags on basic of job definition.
 
 Rest done.
 (I guess.... And I hope sooo.)

jhajduga committed:
Working code .

In new jobs U need to add flags
StoreTracks= true
StorePositions= true
StoreEnergies = true
StorePrimaries = true
StoreRunInfo= true
MinimalMode= false

In RunSvc to run full NtupleAnalisys

jhajduga commented:
I guess its redy for review. 
@barachwal

barachwal commented on app/tld/src/TLD.cc:
```diff
@@ -52,17 +52,16 @@ void TLD::CellVoxelisedScorer(G4bool val) {
-: VPatient(label), m_tld_medium(tldMediumName), m_stl_geometry_file_path(stlGeometryFilePath){
-    m_centre = centre;
+    : VPatient(label), m_tld_medium(tldMediumName), m_stl_geometry_file_path(stlGeometryFilePath) {
+  m_centre = centre;
```
w sumie, to też można przenieść do listy inijcalizacyjnej.

barachwal commented on app/tld/src/TLDTray.cc:
```diff
@@ -174,6 +174,7 @@ std::map<std::size_t, VoxelHit> TLDTray::GetScoringHashedMap(const G4String& sco
               auto voxelCentre = tld_sv->GetVoxelCentre(ix,iy,iz);
             //   G4cout << " voxel "<< ix <<","<<iy<<","<<iz<<" centre: " << voxelCentre << G4endl;
               hashed_map_scoring[voxelHash].SetCentre(voxelCentre);
+              hashed_map_scoring[voxelHash].SetStoreTracks(Service<ConfigSvc>()->GetValue<bool>("RunSvc", "StoreTracks"));
```
Jeśli w kontekście jednej funkcji więcej niż raz korzystamy z wartości trzymanej w configSvc to warto ją przypisać do lokalnej zmiennej, by za każdym razem nie wyciagać jej z serwisu, a poza tym, łatwiej utrzymać kod, bo mniej razy wystepuje `Service<ConfigSvc>()->GetValue<bool>("RunSvc", "StoreTracks")`

barachwal commented on core/analysis/AnalysisFlags.cc:
```diff
@@ -0,0 +1,52 @@
+#include <iostream>
+#include <iomanip>
+
+AnalysisFlagRegistry& AnalysisFlagRegistry::Instance() {
```
Dotychczas konstruowaliśmy singletony zwracając wskaźnik, nazwa metody też inna:
```
RunSvc* RunSvc::GetInstance() {
  static RunSvc instance;
  return &instance;
}
```

barachwal commented on core/analysis/AnalysisFlags.cc:
```diff
@@ -0,0 +1,52 @@
+
+bool AnalysisFlagRegistry::IsEnabled(AnalysisFlag flag) const { return m_flags[flag]; }
+
+void AnalysisFlagRegistry::ResetAll() { m_flags.Reset(); }
```
Jednolinijkowce do pliku .h

barachwal commented on core/analysis/AnalysisFlags.hh:
```diff
@@ -0,0 +1,50 @@
+  void Reset() { flags.reset(); }
+
+ private:
+  std::bitset<static_cast<size_t>(AnalysisFlag::COUNT)> flags;
```
brak konwencji, może tak: m_bitset_flags

barachwal commented on core/analysis/AnalysisFlags.hh:
```diff
@@ -0,0 +1,50 @@
+
+ private:
+  AnalysisFlagRegistry() = default;
+  AnalysisFlags m_flags;
```
proponuję pełniejszą nazwe: m_analysis_flags;

barachwal commented on core/analysis/NTupleEventAnalisys.cc:
```diff
@@ -1,476 +1,502 @@
 ///
-NTupleEventAnalisys *NTupleEventAnalisys::GetInstance() {
+NTupleEventAnalisys* NTupleEventAnalisys::GetInstance() {
   static NTupleEventAnalisys instance = NTupleEventAnalisys();
```
można poprawić, wystarczy:
 static NTupleEventAnalisys instance;

barachwal commented on core/analysis/NTupleEventAnalisys.cc:
```diff
@@ -1,476 +1,502 @@
+
+
+    if (cellVoxelisation) {
+      AnalysisFlagRegistry::Instance().SetFlag(AnalysisFlag::Voxelized, true);
```
aby za każdym razem nie dobierać się w tej klasie do AnalysisFlagRegistry poprzez Instance, można zrobić do niej odwołanie w tej klasie, poprzez atrybut, czyli AnalysisFlagRegistry m_analysis_reg;

barachwal commented on core/analysis/NTupleEventAnalisys.cc:
```diff
@@ -1,476 +1,502 @@
+  SetAnalysisFlag(AnalysisFlag::StoreEnergies, Service<ConfigSvc>()->GetValue<bool>("RunSvc", "StoreEnergies"));
+  SetAnalysisFlag(AnalysisFlag::StoreTracks, Service<ConfigSvc>()->GetValue<bool>("RunSvc", "StoreTracks"));
+  SetAnalysisFlag(AnalysisFlag::StorePrimaries, Service<ConfigSvc>()->GetValue<bool>("RunSvc", "StorePrimaries"));
+  SetAnalysisFlag(AnalysisFlag::MinimalMode, Service<ConfigSvc>()->GetValue<bool>("RunSvc", "MinimalMode"));
```
Ta metoda wykorzystana jest tylko tutaj, wiec moze nie trzeba z niej robić składowej klasy? Wystarczyło by ja tutaj mieć jako lambde.

barachwal commented on core/geometry/Patient/D3DDetector/D3DDetector.cc:
```diff
@@ -464,80 +445,74 @@ std::map<std::size_t, VoxelHit> D3DDetector::GetScoringHashedMap(const G4String&
-G4bool D3DDetector::IsInside(double x, double y, double z) {
-  return false;
-}
+G4bool D3DDetector::IsInside(double x, double y, double z) { return false; }
```
jednolinijkowiec, do pliku .h

barachwal commented on core/geometry/Patient/VoxelHit.cc:
```diff
@@ -47,239 +49,388 @@ void VoxelHit::Fill(G4Step* aStep){
 ////////////////////////////////////////////////////////////////////////////////
 ///
-G4int VoxelHit::GetTrkType(G4Step* aStep) const {
+G4int VoxelHit::GetTrackIdTypeMappingList(G4Step* aStep) const {
```
Chyba poleciałeś z automatu zamieniając nazwe tej metody również... chcemy to jednak rozróżnić, wiec mamy dwie metody:
```
std::vector<std::pair<G4int, G4int>> GetTrackIdTypeMappingList() const;
```
oraz 
```
G4int GetTrackIdTypeMappingList(G4Step* aStep) const;
```
Wiec tutaj, dla metody która zwraca pojedyńczy trkType, cofnijmy zmiane nazwy, albo zróby odpowiednio do nowego: `G4int GetTrackIdType(G4Step* aStep) const;`

jhajduga committed:
Changes from comment in PR done.

jhajduga commented:
All the changes proposed in the comments in review have been implemented.
